### DISCUSSION TITLE
Dump a body ( and headers ) of the HTTP/S POST/PATCH request

### DISCUSSION THREAD

**Discussion by myers-dev** [canton7/RestEase#221] [Q&A]
Can I dump a body and a headers of the POST/PATCH request before the request will be sent out ? 

For sure, I can dump a class object but, for troubleshooting purposes, I'm looking to see a _raw_ body / headers . 
`            Console.WriteLine(JsonConvert.SerializeObject(gitPush));
            Response<GitPush> resp7 = api7.GitPushCreate(gitPush).Result;`

Thank you !

**Comment by canton7** [comment]
The easiest thing is probably to grab Fiddler, and see the request actually go out over the network.

Alternatively, you can use a [`RequestModifier`](https://github.com/canton7/RestEase#requestmodifier) to dump the request, or use a custom `HttpClientHandler` or [`LoggingHttpMessageHandler`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.http.logging.logginghttpmessagehandler?view=dotnet-plat-ext-6.0).

**Reply by myers-dev** [reply]
Thank you @canton7 for a quick reply ! I used Fiddler and successfully captured the body. 

Would it be possible to suppress **null** attributes from being included in the JSON ?
so , instead of JSON like below , I'd got shorter version ?

`{"_links":null,"Commits":[{"_links":null,"Author":null,"ChangeCounts":null,"Changes":[{"ChangeId":null,"ChangeType":"add","Item":{"Path":"/file.txt"},"NewContent":{"Content":"Hello World !","ContentType":"rawtext"},"NewContentTemplate":null,"OriginalPath":null,"SourceServerItem":null,"Url":null}],"Comment":"Initial commit for feature3","CommentTruncated":null,"CommitId":null,"Committer":null,"Parents":null,"Push":null,"RemoteUrl":null,"Statuses":null,"Url":null,"WorkItems":null}],"Date":null,"PushId":null,"PushedBy":null,"RefUpdates":[{"IsLocked":null,"Name":"refs/heads/feature3","NewObjectId":null,"OldObjectId":"0000000000000000000000000000000000000000","RepositoryId":null}],"Repository":null,"Url":null}`

**Reply by canton7** [reply]
Glad that worked! Fiddler's a generally awesome tool for this sort of thing.

> Would it be possible to suppress null attributes from being included in the JSON ?

That's a json.net thing. See e.g. [this StackOverflow thread](https://stackoverflow.com/questions/6507889/how-to-ignore-a-property-in-class-if-null-using-json-net).

**Reply by myers-dev** [reply]
Thank you ! Just to confirm, there is no way to do it with RestEase ? ( e.g when class is serialized to Json body, not to include the null fields ) ?

**Reply by canton7** [reply]
RestEase uses json.net for JSON serialization. By default, it just calls `JsonConvert.SerializeObject`, [see here](https://github.com/canton7/RestEase/blob/c68094d3c4c731eda570f0fd44337b03cd213ade/src/RestEase/JsonRequestBodySerializer.cs). So, just use the json.net attributes / configuration, and this will get used by json.net when RestEase calls into it to do the seralization.

**Reply by myers-dev** [reply]
Thank you. 

I'm using Newton.Json and the syntax is similar. I updated the class as follows ( for all attributes ) . It solved my problem.

```
    public class GitCommitRef
    {
        [JsonProperty(PropertyName = "_links", NullValueHandling = NullValueHandling.Ignore)]
        public ReferenceLinks? Links { get; set; }
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
        public GitUserDate? Author { get; set; }
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
        public object? ChangeCounts { get; set; }
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
```

To note, when body JSON had attributes with null value, the Azure Devops throwed the following error 

```
{'$id': '1', 'innerException': None, 'message': 'The body of the request contains invalid Json.\r\nParameter name: contentStream', 'typeName': 'Microsoft.TeamFoundation.SourceControl.WebServer.InvalidArgumentValueException, Microsoft.TeamFoundation.SourceControl.WebServer', 'typeKey': 'InvalidArgumentValueException', 'errorCode': 0, 'eventId': 0}
```

**Reply by canton7** [reply]
Newtonsoft.json and json.net are two names for the same thing.
